autobot-backend/knowledge/fact_projection.py

```
@@ -0,0 +1,186 @@
+    async def _read_fact_for_write(self, fact_id: str) -> tuple[Dict[str, Any], Dict[str, Any]] | None:
+        """The fact as it stands, for an update or a delete. ``None`` if unknown.
+
+        Reads the Redis projection first because it is warm, and falls back to
+        the durable row — which is the case #12733 could not survive: the hashes
+        were gone, so every fact "did not exist" even though the knowledge was
+        still recoverable. The row settles existence; Redis only makes it fast.
+        """
+        # Lazy: see the module docstring on why fact_store is not imported at module scope.
+        from knowledge import fact_store
+
+        fact_data = await asyncio.to_thread(self.redis_client.hgetall, "fact:%s" % fact_id)
```

Ruff (0.16.3)

[warning] Use format specifiers instead of percent format

Replace with format specifiers

(UP031)